### PR TITLE
fix data mask for oracle

### LINKED ISSUE #3194: Data masking or encryption for Oracle can not work well.
Please see this thread for more detail. 
This thread explains the reason and gives the solution. 
If anyone is interested in this bugfix, please reply this thread or this issue. Hope see your Pr later. 😃 
> https://lists.apache.org/thread.html/b040322a2bbddd4994ab0829b85ed79b94036e5d82716d81993d0d4f@%3Cdev.shardingsphere.apache.org%3E

### PR DESCRIPTION
**Pull request by yanyzy**
Fixes #3194.

add getTableName for oracle.

### CODE REVIEW

File: sharding-core/sharding-core-preprocessor/src/main/java/org/apache/shardingsphere/core/preprocessor/statement/impl/SelectSQLStatementContext.java [hunk 1]
```diff
@@ -177,4 +181,12 @@ private String getOrderItemText(final TextOrderByItemSegment orderByItemSegment)
     public boolean isSameGroupByAndOrderByItems() {
         return !groupByContext.getItems().isEmpty() && groupByContext.getItems().equals(orderByContext.getItems());
     }
+
+    /**
+     * the table metadata.
+     * @return table metadata
+     */
+    public TableMetas tableMetas() {
```
Comment (tristaZero) [inline sharding-core/sharding-core-preprocessor/src/main/java/org/apache/shardingsphere/core/preprocessor/statement/impl/SelectSQLStatementContext.java]: What's tableMetas() used for?

File: sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/communication/jdbc/execute/callback/ProxySQLExecuteCallback.java [hunk 2]
```diff
@@ -61,14 +62,17 @@
     private final boolean fetchMetaData;
     
     private boolean hasMetaData;
+
+    private SQLStatementContext sqlStatementContext;
     
     public ProxySQLExecuteCallback(final BackendConnection backendConnection, final JDBCExecutorWrapper jdbcExecutorWrapper, 
-                                   final boolean isExceptionThrown, final boolean isReturnGeneratedKeys, final boolean fetchMetaData) {
+                                   final boolean isExceptionThrown, final boolean isReturnGeneratedKeys, final boolean fetchMetaData, final SQLStatementContext sqlStatementContext) {
```
Comment (tristaZero) [inline sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/communication/jdbc/execute/callback/ProxySQLExecuteCallback.java]: Considering that the type of isExceptionThrown, isReturnGeneratedKeys and fetchMetaData is boolean, do you think it is better if we put sqlStatementContext in front of those paramters?

File: sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultMetaData.java [hunk 3]
```diff
@@ -155,6 +171,29 @@ private String getTableName(final int columnIndex) throws SQLException {
...
+    private boolean isCurrentTable(final int columnIndex, final String logicTableName) throws SQLException {
+        Collection<String> actualColumns = encryptRule.findEncryptTable(logicTableName)
+                .transform(new Function<EncryptTable, Collection<String>>() {
+                    @Override
+                    public Collection<String> apply(final EncryptTable encryptTable) {
+                        return encryptTable.getCipherColumns();
+                    }
+                })
+                .or(Collections.<String>emptyList());
+        String actualColumn = resultSetMetaData.getColumnName(columnIndex);
```
Comment (tristaZero) [inline sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultMetaData.java]: Hi, to get actualColumn, you use `resultSetMetaData.getColumnName(columnIndex)` which is called in `getLogicColumn(final String tableName, final int columnIndex)` as well.
**Notice** The purpose of getShardingEncryptor(final int columnIndex) is to get Optional<ShardingEncryptor>, so i think you can merge `isCurrentTable()` and `getLogicColumn(final String tableName togather to get the final result, i.e, Optional<ShardingEncryptor>.

File: sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultMetaData.java [hunk 4]
```diff
@@ -144,6 +157,9 @@ public boolean isCaseSensitive(final int columnIndex) throws SQLException {
      */
     public Optional<ShardingEncryptor> getShardingEncryptor(final int columnIndex) throws SQLException {
         String logicTable = getTableName(columnIndex);
```
Comment (tristaZero) [inline sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultMetaData.java]: If the new function, `getTableName(columnIndex, sqlStatementContext)` is availiable, do you think this function is unneeded?

File: sharding-core/sharding-core-preprocessor/src/main/java/org/apache/shardingsphere/core/preprocessor/statement/impl/SelectSQLStatementContext.java (hunk 1 above)
Comment (yanyzy) [reply sharding-core/sharding-core-preprocessor/src/main/java/org/apache/shardingsphere/core/preprocessor/statement/impl/SelectSQLStatementContext.java]: Sorry, i forgot to remove it.

File: sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultMetaData.java (hunk 3 above)
Comment (yanyzy) [reply sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultMetaData.java]: Good idea, i will modify it like this!

File: sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultMetaData.java [hunk 5]
```diff
@@ -143,21 +149,35 @@ public boolean isCaseSensitive(final int columnIndex) throws SQLException {
...
-        String columnLabel = resultSetMetaData.getColumnName(columnIndex);
-        return encryptRule.isCipherColumn(tableName, columnLabel) ? encryptRule.getLogicColumn(tableName, columnLabel) : columnLabel;
+
+    private Optional<ShardingEncryptor> findShardingEncryptorWithTable(final String actualColumn, final String logicTableName) {
+        if (null == encryptRule) {
+            return Optional.absent();
+        }
+        final Collection<String> actualColumns = encryptRule.findEncryptTable(logicTableName)
+                .transform(new Function<EncryptTable, Collection<String>>() {
+                    @Override
```
Comment (tristaZero) [inline sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultMetaData.java]: Please add blank line before @override, which is just checkstyle.

**Comment by tristaZero** [comment]
I approve it after modifying the checkstyple, how about you? @terrymanu 
@yanyzy Thanks for your contribution, it is great. 👍

File: sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultMetaData.java (hunk 5 above)
Comment (yanyzy) [reply sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultMetaData.java]: Done.

**Comment by tristaZero** [comment]
@yanyzy Well done :)

**Pull request merged**